データベース設計

keeee-a opened the pull request:
# What
データベースの設計

# Why
データベース作成の見通しをするため

[ER図](https://gyazo.com/f336b2f191f34381faf026a450b1b503)

keeee-a committed:
データベース設計

chirikuzu commented on README.md:
```diff
@@ -1,24 +1,25 @@
+
+|Column       |Type         |Options       |
+|-------------|-------------|--------------|
+|name         |string       |null:false    |
```
```
|name         |string       |null:false    |
```
ユーザーの登録時、姓、名、姓（フリガナ）、名（フリガナ）と入力欄が細かく分かれているかと思われます。
カラムは姓、名、姓（フリガナ）、名（フリガナ）とそれぞれ設計しましょう。

chirikuzu commented on README.md:
```diff
@@ -1,24 +1,25 @@
+|Column       |Type         |Options       |
+|-------------|-------------|--------------|
+|name         |string       |null:false    |
+|e-mail       |string       |null:false    |
```
```
|e-mail       |string       |null:false    |
```
今回ユーザー管理機能にはdeviseを使用して実装を行っていきますので、メールアドレスのカラムのカラム名はdevise側がデフォルトで用意する`email`にしておきましょう。

chirikuzu commented on README.md:
```diff
@@ -1,24 +1,25 @@
+|name         |string       |null:false    |
+|e-mail       |string       |null:false    |
+|nickname     |string       |null:false    |
+|password     |string       |null:false    |
```
```
|password     |string       |null:false    |
```
passwordのカラム名はdevise側がデフォルトで用意している`encrypted_password`としましょう。

deviseを使用する場合、新規登録の際にpasswordというキーに対して入力された値は暗号化されてencrypted_passwordというカラムに保存されます。
そのため、このencrypted_passwordというカラムをpasswordカラムに変更してしまうとエラーが発生してしまいます。
DB設計の段階からdeviseの仕様に合わせておくことで、エラーを未然に防ぐことができます。

chirikuzu commented on README.md:
```diff
@@ -1,24 +1,25 @@
+|Column           |Type            |Options       |
+|-----------------|----------------|--------------|
+|name             |string          |null:false    |
+|category         |string          |null:false    |
```
```
|category         |string          |null:false    |
```
カテゴリーのカラムはactive_hashで実装の予定のため、integer型にしましょう。
また、カラム名にはactive_hashに対応させるため`_id`をつけましょう。
詳しくは「カテゴリーの選択を工夫しよう」を参考にしてみてください。

chirikuzu commented on README.md:
```diff
@@ -1,24 +1,25 @@
+|name             |string          |null:false    |
+|category         |string          |null:false    |
+|price            |integer         |null:false    |
+|image            |                |null:false    |
```
```
|image            |                |null:false    |
```
こちらのカラムですが、今回の実装では商品出品の画像投稿において「active_storage」を使用するため、設計の段階から削除しておきましょう。
理由といたしましては、active_storage導入時に自動でテーブルなどが生成されるからです。

chirikuzu commented on README.md:
```diff
@@ -1,24 +1,25 @@
+|price            |integer         |null:false    |
+|image            |                |null:false    |
+|explanation      |text            |null:false    |
+|product_condition|text            |null:false    |
+|burden_shipping  |text            |null:false    |
+|shipping_area    |text            |null:false    |
+|days_to_ship     |integer         |null:false    |
```
```
|product_condition|text            |null:false    |
|burden_shipping  |text            |null:false    |
|shipping_area    |text            |null:false    |
|days_to_ship     |integer         |null:false    |
```
商品の状態, 配送料の負担, 発送元の地域,  発送までの日数のカラムについてもactive_hashで実装の予定のため、integer型にしましょう。
また、カラム名にはactive_hashに対応させるため`_id`をつけましょう。
詳しくは「カテゴリーの選択を工夫しよう」を参考にしてみてください。

chirikuzu commented on README.md:
```diff
@@ -1,24 +1,25 @@
-* Deployment instructions
-
-* ...
+# テーブル設計
```
DB設計全体について2点コメントです。

- ユーザーは新規登録、出品の他に購入を行いますのでそれに関わるテーブルを設計しましょう。
詳しくは**実装に悩んだときのヒント：データベース設計**のカリキュラムを参考にしてみましょう。

- 各テーブル同士のアソシエーションの設計を行いましょう。

chirikuzu reviewed (commented):
レビュー依頼ありがとうございます。
コメントしましたので、ご確認をお願いいたします。

keeee-a committed:
データベース設計修正1

設計テーブルの修正
purchasesテーブル、addressesテーブルの作成
associationの記述

keeee-a commented:
データベース設計修正しました！
[ER図](https://gyazo.com/96bbc7901ddddd7adb4754cea9bcab80)

shodai-ito commented on README.md:
```diff
@@ -1,24 +1,57 @@
+|last_name           |string       |null:false    |
+|first_name_kana     |string       |null:false    |
+|last_name_kana      |string       |null:false    |
+|email               |string       |null:false    |
```
こちらemailには一意性制約をつけましょう！

shodai-ito commented on README.md:
```diff
@@ -1,24 +1,57 @@
+## purchases テーブル
+|Column               |Type            |Options           |
+|---------------------|----------------|------------------|
+|user_id              |integer         |foreign_key: true |
```
`foreign_key: true`
というオプションは、reference型に使用されるオプションですので、データ型を変更しましょう！

shodai-ito commented on README.md:
```diff
@@ -1,24 +1,57 @@
+|---------------------|----------------|-------------------|
+|postal_code_id       |integer         |null:false         |
+|prefecture_id        |integer         |null:false         |
+|city_name_id         |integer         |null:false         |
```
こちら、カラムが不足しているかと思われます。
サンプルアプリを参考に、カラムを追加しましょう！
また、郵便番号や市区町村はアクティブハッシュを利用しないので、id不要です！

shodai-ito commented on README.md:
```diff
@@ -1,24 +1,57 @@
+|city_name_id         |integer         |null:false         |
+
+### Association 
+belongs_to :user
```
こちらのアソシエーションでも間違いではないのですが、
今回は見本アプリの設計上、「配送先住所と購入管理が紐づく」実装を行なっていただく必要がございます。
よって、ユーザーと住所のアソシエーションは組まなくてもOKです！

shodai-ito reviewed (commented):
レビュー依頼ありがとうございます！
追加で数点コメントさせていただきました。
ご確認をお願いします！

keeee-a committed:
データベース設計修正2

emailカラムに一意性制約付与
purchaseテーブルの修正
street_addressesテーブルのカラム追加、option変更、association消去

keeee-a commented:
二度目の修正を行いました！
よろしくお願いします！

Akinori0123 commented on README.md:
```diff
@@ -1,24 +1,57 @@
+|birthday            |date         |null:false                 |
+
+### Association
+has_one :street_address
```
```
has_one :street_address
```
前回レビューにありました通り、ユーザと住所のアソシエーションは不要となります。
上記記述も削除しましょう。

Akinori0123 commented on README.md:
```diff
@@ -1,24 +1,57 @@
+|burden_shipping_id   |integer         |null:false    |
+|shipping_area_id     |integer         |null:false    |
+|days_to_ship_id      |integer         |null:false    |
+|user_id              |integer         |null:false    |
```
```
|user_id              |integer         |null:false    |
```
外部キーとして扱うため外部キー制約をつけましょう（`foreign_key: true`）。
前回レビューにありました通り、データ型をintegerからreferencesに変更していただければと思います。
references型ではカラム名に「_id」は不要になるため、カラム名も修正しましょう。

全てのテーブルの外部キーカラムに対して同様に修正しましょう。

Akinori0123 commented on README.md:
```diff
@@ -1,24 +1,57 @@
+### Association
+belongs_to :product
+belongs_to :user
+
```
**purchasesのアソシエーション**
「購入履歴は一つ住所を持っている」といったアソシエーションを追加しましょう。
street_addressesのアソシエーションにも購入履歴とのアソシエーションを追加しましょう。

Akinori0123 commented on README.md:
```diff
@@ -1,24 +1,57 @@
+|postal_code          |integer         |null:false         |
+|prefecture_id        |integer         |null:false         |
+|city                 |string          |null:false         |
+|address              |string          |null:false 
```
**street_addresses テーブル**
建物名、電話番号、purchasesテーブルに繋がる外部キーの三つのカラムを追加しましょう。

Akinori0123 reviewed (commented):
レビュー依頼ありがとうございます。前回の修正確認いたしました。
追加でコメントしましたのでご確認いただき、修正後改めてレビュー依頼をいただけますと幸いです。よろしくお願いいたします。

keeee-a committed:
データベース設計修正3

・ユーザと住所のアソシエーションの消去
・外部キーカラムに対するreferencesへのデータ型変更と、カラム名変更
・purchasesテーブルの購入履歴とのアソシエーション設定
・street_addressesテーブルに建物名、電話番号
、外部キーカラムを追加

keeee-a commented:
修正確認ありがとうございました！
修正しました。

Akinori0123 commented on README.md:
```diff
@@ -1,24 +1,59 @@
+## street_addresses テーブル
+|Column               |Type            |                              |
+|---------------------|----------------|------------------------------|
+|postal_code          |integer         |null:false                    |
```
【修正後のレビュー依頼は不要となります】
```
|postal_code          |integer         |null:false                    |
```
integer型で実装するアプリもありますが、今回のフリマアプリではハイフン（文字列）を含むため郵便番号はstring型にしましょう。

Akinori0123 commented on README.md:
```diff
@@ -1,24 +1,59 @@
+|city                 |string          |null:false                    |
+|address              |string          |null:false                    |
+|building             |string          |                              |
+|phone_number         |integer         |null:false                    |
```
【修正後のレビュー依頼は不要となります】
```
|phone_number         |integer         |null:false                    |
```
電話番号は数字の文字列として実装するためstring型が適切となります。
（integer型にすると電話番号の最初の"0"がデータベースに保存できなくなってしまいます。詳しくは「電話番号 string integer」でご検索いただければと思います）

Akinori0123 reviewed (commented):
レビュー依頼ありがとうございます。
大枠のテーブルとカラムについてLGTMです。型についてコメントをしましたのでマージ前にご確認お願いいたします。
実装お疲れ様でした。引き続き頑張っていきましょう！

keeee-a committed:
データベース設計修正4

細かいところもろもろ
